src/search_controller.ts

```diff
@@ -0,0 +1,484 @@
+    const { debounceMs, isActive, pageSize } = { ...DEFAULT_SEARCH_SOURCE_OPTIONS, ...options };
+    this.pageSize = pageSize;
+    this.state = new StateStore<SearchSourceState<T>>({
+      hasMore: true,
+      isActive,
+      isLoading: false,
+      items: undefined,
+      offset: 0,
+      searchQuery: '',
+    });
+    this.setDebounceOptions({ debounceMs });
+  }
+
+  get lastQueryError() {
+    return this.state.getLatestValue().lastQueryError;
+  }
+
+  get hasMore() {
+    return this.state.getLatestValue().hasMore;
+  }
+
+  get hasResults() {
+    return Array.isArray(this.state.getLatestValue().items);
+  }
+
+  get isActive() {
+    return this.state.getLatestValue().isActive;
+  }
+
+  get isLoading() {
+    return this.state.getLatestValue().isLoading;
+  }
+
+  get items() {
+    return this.state.getLatestValue().items;
+  }
+
+  get next() {
+    return this.state.getLatestValue().next;
+  }
+
+  get offset() {
+    return this.state.getLatestValue().offset;
+  }
+
+  get searchQuery() {
+    return this.state.getLatestValue().searchQuery;
+  }
+
+  protected abstract query(searchQuery: string): Promise<QueryReturnValue<T>>;
+
+  protected abstract filterQueryResults(items: T[]): T[] | Promise<T[]>;
+
+  setDebounceOptions = ({ debounceMs }: DebounceOptions) => {
+    this.searchDebounced = debounce(this.executeQuery.bind(this), debounceMs);
+  };
+
+  activate = (sourceStateOverride?: Partial<SearchSourceState>) => {
+    if (this.isActive) return;
+    if (this.searchQuery) {
+      this.search();
+    }
+    this.state.partialNext({ ...sourceStateOverride, isActive: true });
+  };
+
+  deactivate = () => {
+    if (!this.isActive) return;
+    this.state.partialNext({ isActive: false });
+  };
+
+  async executeQuery(newSearchString?: string) {
+    const hasNewSearchQuery = typeof newSearchString !== 'undefined';
+    const searchString = newSearchString ?? this.searchQuery;
+    if (!this.isActive || this.isLoading || (!this.hasMore && !hasNewSearchQuery) || !searchString) return;
+
+    if (hasNewSearchQuery) {
+      this.resetState({ isActive: this.isActive, isLoading: true, searchQuery: newSearchString });
+    } else {
+      this.state.partialNext({ isLoading: true });
+    }
+
+    const stateUpdate: Partial<SearchSourceState<T>> = {};
+    try {
+      const results = await this.query(searchString);
+      if (!results) return;
+      const { items, next } = results;
+
+      if (next) {
+        stateUpdate.next = next;
+        stateUpdate.hasMore = !!next;
+      } else {
+        stateUpdate.offset = (this.offset ?? 0) + items.length;
+        stateUpdate.hasMore = items.length === this.pageSize;
+      }
+
+      stateUpdate.items = await this.filterQueryResults(items);
+    } catch (e) {
+      stateUpdate.lastQueryError = e as Error;
+    } finally {
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      this.state.next(({ lastQueryError, ...prev }: SearchSourceState<T>) => ({
```

Comment:
```suggestion
      this.state.next((current) => ({
```
Also, `stateUpdate` seems to get unwrapped after the "prev" (current state) and unless it (`stateUpdate`) contains reset value it does not really seem necessary to exclude it (reset value is not being set in the try block).

Reply:
The reason for leaving the `lastQueryError` from the previous state out of the next update is to prevent carrying over errors relevant to previous query if the current query hasn't end up in an error. If you have done 2 queries, in the second query you stored the error, then if the 3rd query has not end up in an error, you do not want to keep the error related to the 2nd query.